dockerspawner/dockerspawner.py

```diff
@@ -360,10 +372,20 @@ def start(self, image=None, extra_create_kwargs=None,
         # start the container
         yield self.docker('start', self.container_id, **start_kwargs)
 
-        # get the public-facing ip, port
-        resp = yield self.docker('port', self.container_id, 8888)
-        self.user.server.ip = self.container_ip
-        self.user.server.port = resp[0]['HostPort']
+        ip, port = yield from self.get_ip_and_port()
+        self.user.server.ip = ip
+        self.user.server.port = port
+
+    def get_ip_and_port(self):
+        if self.use_internal_ip:
+            inspect_resp = yield self.docker('inspect_container', self.container_id)
+            ip = inspect_resp['NetworkSettings']['IPAddress']
```

Comment:
Not sure if you get notified if I edit my original post so here as separate post:

Tests done. unfortunately, even if the container is not removed but just stopped, it can happen that the IP address is reassigned. This means if the user manually stops containers or removes them the IP address might be outdated, so I don't think caching is an option here. Let me know how you want to proceed.